File: test_lambda_memory.py
"""
Script para probar el Lambda con AgentCore Memory
Ejecutar: python test_lambda_memory.py
"""
import boto3
import json
import time

# Configuración
LAMBDA_NAME = "gpbible-bedrock-processor-memory-test"
REGION = "us-east-1"
MEMORY_ID = "memory_bqdqb-jtj3lc48bl"

lambda_client = boto3.client('lambda', region_name=REGION)
agentcore_client = boto3.client('bedrock-agentcore', region_name=REGION)
logs_client = boto3.client('logs', region_name=REGION)


def invoke_lambda(user_id: str, conversation_id: str, message_id: str, text: str):
    """Invoca el Lambda con un mensaje de prueba"""
    
    payload = {
        "Records": [{
            "Sns": {
                "Message": json.dumps({
                    "conversationId": conversation_id,
                    "messageId": message_id,
                    "userId": user_id,
                    "text": text
                })
            }
        }]
    }
    
    print(f"\n📤 Enviando: {text[:50]}...")
    
    response = lambda_client.invoke(
        FunctionName=LAMBDA_NAME,
        InvocationType='RequestResponse',
        Payload=json.dumps(payload)
    )
    
    result = json.loads(response['Payload'].read())
    
    if 'FunctionError' in response:
        print(f"⚠️  Lambda ejecutado con error (esperado si webhook no está configurado)")
    else:
        print(f"✅ Lambda ejecutado exitosamente")
    
    return result


def check_memory(user_id: str):
    """Verifica la memoria guardada para un usuario"""
    
    print(f"\n🔍 Verificando memoria para: {user_id}")
    
    try:
        # Listar sesiones
        sessions = agentcore_client.list_sessions(
            memoryId=MEMORY_ID,
            actorId=user_id
        )
        
        session_list = sessions.get('sessionSummaries', [])
        print(f"   Sesiones encontradas: {len(session_list)}")
        
        for session in session_list:
            session_id = session['sessionId']
            print(f"\n   📁 Sesión: {session_id}")
            
            # Listar eventos de la sesión
            events = agentcore_client.list_events(
                memoryId=MEMORY_ID,
                actorId=user_id,
                sessionId=session_id,
                maxResults=10
            )
            
            for event in events.get('events', []):
                payload = event.get('payload', [])
                if payload and payload[0].get('conversational'):
                    conv = payload[0]['conversational']
                    role = conv.get('role', 'UNKNOWN')
                    text = conv.get('content', {}).get('text', '')[:100]
                    print(f"      {role}: {text}...")
        
        return len(session_list)
        
    except Exception as e:
        print(f"   Error: {e}")
        return 0


def get_recent_logs(limit: int = 10):
    """Obtiene los logs más recientes del Lambda"""
    
    log_group = f"/aws/lambda/{LAMBDA_NAME}"
    
    try:
        streams = logs_client.describe_log_streams(
            logGroupName=log_group,
            orderBy='LastEventTime',
            descending=True,
            limit=1
        )
        
        if not streams['logStreams']:
            print("No hay logs disponibles")
            return
        
        stream_name = streams['logStreams'][0]['logStreamName']
        
        events = logs_client.get_log_events(
            logGroupName=log_group,
            logStreamName=stream_name,
            limit=limit
        )
        
        print(f"\n📋 Últimos logs ({stream_name}):")
        print("-" * 60)
        
        for event in events['events']:
            msg = event['message'].strip()
            if any(x in msg for x in ['INFO', 'ERROR', 'Memory', 'memory', 'Fetching', 'saved', 'Found']):
                # Limpiar mensaje
                if '\t' in msg:
                    parts = msg.split('\t')
                    if len(parts) >= 3:
                        msg = parts[2]
                print(msg[:200])
        
    except Exception as e:
        print(f"Error obteniendo logs: {e}")


def run_test():
    """Ejecuta una prueba completa"""
    
    print("=" * 60)
    print("🧪 PRUEBA DE LAMBDA CON AGENTCORE MEMORY")
    print("=" * 60)
    
    # Usar un userId único para la prueba
    test_user = f"test-user-{int(time.time())}"
    
    # 1. Primera conversación (sin memoria previa)
    print("\n" + "=" * 60)
    print("1️⃣  PRIMERA CONVERSACIÓN (sin memoria previa)")
    print("=" * 60)
    
    invoke_lambda(
        user_id=test_user,
        conversation_id="conv-001",
        message_id="msg-001",
        text="Hola, estoy pasando por un momento difícil y necesito guía espiritual"
    )
    
    time.sleep(2)
    
    # 2. Verificar que se guardó la memoria
    print("\n" + "=" * 60)
    print("2️⃣  VERIFICANDO MEMORIA GUARDADA")
    print("=" * 60)
    
    sessions = check_memory(test_user)
    
    if sessions == 0:
        print("❌ No se guardó la memoria")
        return
    
    # 3. Segunda conversación (debería leer memoria)
    print("\n" + "=" * 60)
    print("3️⃣  SEGUNDA CONVERSACIÓN (con memoria)")
    print("=" * 60)
    
    invoke_lambda(
        user_id=test_user,
        conversation_id="conv-002",
        message_id="msg-002",
        text="¿Qué versículo me recomiendas para encontrar paz?"
    )
    
    time.sleep(2)
    
    # 4. Ver logs
    print("\n" + "=" * 60)
    print("4️⃣  LOGS DEL LAMBDA")
    print("=" * 60)
    
    get_recent_logs(15)
    
    # 5. Verificar memoria final
    print("\n" + "=" * 60)
    print("5️⃣  MEMORIA FINAL")
    print("=" * 60)
    
    check_memory(test_user)
    
    print("\n" + "=" * 60)
    print("✅ PRUEBA COMPLETADA")
    print("=" * 60)
    print(f"\nUsuario de prueba: {test_user}")
    print("Revisa los logs para ver '✅ Found previous conversations' en la segunda llamada")


if __name__ == "__main__":
    run_test()

File: test_lambda_memory.js
/**
 * Script para probar el Lambda con AgentCore Memory
 * Ejecutar: node test_lambda_memory.js
 */

const { LambdaClient, InvokeCommand } = require("@aws-sdk/client-lambda");
const { BedrockAgentCoreClient, ListSessionsCommand, ListEventsCommand } = require("@aws-sdk/client-bedrock-agentcore");
const { CloudWatchLogsClient, DescribeLogStreamsCommand, GetLogEventsCommand } = require("@aws-sdk/client-cloudwatch-logs");

// Configuración
const LAMBDA_NAME = "gpbible-bedrock-processor-memory-test";
const REGION = "us-east-1";
const MEMORY_ID = "memory_bqdqb-jtj3lc48bl";

// Clientes AWS
const lambdaClient = new LambdaClient({ region: REGION });
const agentcoreClient = new BedrockAgentCoreClient({ region: REGION });
const logsClient = new CloudWatchLogsClient({ region: REGION });

/**
 * Invoca el Lambda con un mensaje de prueba
 */
async function invokeLambda(userId, conversationId, messageId, text) {
    const payload = {
        Records: [{
            Sns: {
                Message: JSON.stringify({
                    conversationId,
                    messageId,
                    userId,
                    text
                })
            }
        }]
    };

    console.log(`\n📤 Enviando: ${text.substring(0, 50)}...`);

    try {
        const command = new InvokeCommand({
            FunctionName: LAMBDA_NAME,
            InvocationType: "RequestResponse",
            Payload: JSON.stringify(payload)
        });

        const response = await lambdaClient.send(command);
        const result = JSON.parse(new TextDecoder().decode(response.Payload));

        if (response.FunctionError) {
            console.log("⚠️  Lambda ejecutado con error (esperado si webhook no está configurado)");
        } else {
            console.log("✅ Lambda ejecutado exitosamente");
        }

        return result;
    } catch (error) {
        console.error("❌ Error invocando Lambda:", error.message);
        throw error;
    }
}

/**
 * Verifica la memoria guardada para un usuario
 */
async function checkMemory(userId) {
    console.log(`\n🔍 Verificando memoria para: ${userId}`);

    try {
        // Listar sesiones
        const sessionsCommand = new ListSessionsCommand({
            memoryId: MEMORY_ID,
            actorId: userId
        });

        const sessionsResponse = await agentcoreClient.send(sessionsCommand);
        const sessionList = sessionsResponse.sessionSummaries || [];

        console.log(`   Sesiones encontradas: ${sessionList.length}`);

        for (const session of sessionList) {
            const sessionId = session.sessionId;
            console.log(`\n   📁 Sesión: ${sessionId}`);

            // Listar eventos de la sesión
            const eventsCommand = new ListEventsCommand({
                memoryId: MEMORY_ID,
                actorId: userId,
                sessionId: sessionId,
                maxResults: 10
            });

            const eventsResponse = await agentcoreClient.send(eventsCommand);
            const events = eventsResponse.events || [];

            for (const event of events) {
                const payload = event.payload || [];
                if (payload.length > 0 && payload[0].conversational) {
                    const conv = payload[0].conversational;
                    const role = conv.role || "UNKNOWN";
                    const text = (conv.content?.text || "").substring(0, 100);
                    console.log(`      ${role}: ${text}...`);
                }
            }
        }

        return sessionList.length;
    } catch (error) {
        console.log(`   Error: ${error.message}`);
        return 0;
    }
}

/**
 * Obtiene los logs más recientes del Lambda
 */
async function getRecentLogs(limit = 10) {
    const logGroup = `/aws/lambda/${LAMBDA_NAME}`;

    try {
        // Obtener el stream más reciente
        const streamsCommand = new DescribeLogStreamsCommand({
            logGroupName: logGroup,
            orderBy: "LastEventTime",
            descending: true,
            limit: 1
        });

        const streamsResponse = await logsClient.send(streamsCommand);

        if (!streamsResponse.logStreams || streamsResponse.logStreams.length === 0) {
            console.log("No hay logs disponibles");
            return;
        }

        const streamName = streamsResponse.logStreams[0].logStreamName;

        // Obtener eventos del stream
        const eventsCommand = new GetLogEventsCommand({
            logGroupName: logGroup,
            logStreamName: streamName,
            limit: limit
        });

        const eventsResponse = await logsClient.send(eventsCommand);

        console.log(`\n📋 Últimos logs (${streamName}):`);
        console.log("-".repeat(60));

        for (const event of eventsResponse.events || []) {
            let msg = event.message.trim();
            const keywords = ["INFO", "ERROR", "Memory", "memory", "Fetching", "saved", "Found"];
            
            if (keywords.some(kw => msg.includes(kw))) {
                // Limpiar mensaje
                if (msg.includes("\t")) {
                    const parts = msg.split("\t");
                    if (parts.length >= 3) {
                        msg = parts[2];
                    }
                }
                console.log(msg.substring(0, 200));
            }
        }
    } catch (error) {
        console.log(`Error obteniendo logs: ${error.message}`);
    }
}

/**
 * Espera un tiempo determinado
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Ejecuta una prueba completa
 */
async function runTest() {
    console.log("=".repeat(60));
    console.log("🧪 PRUEBA DE LAMBDA CON AGENTCORE MEMORY");
    console.log("=".repeat(60));

    // Usar un userId único para la prueba
    const testUser = `test-user-${Date.now()}`;

    // 1. Primera conversación (sin memoria previa)
    console.log("\n" + "=".repeat(60));
    console.log("1️⃣  PRIMERA CONVERSACIÓN (sin memoria previa)");
    console.log("=".repeat(60));

    await invokeLambda(
        testUser,
        "conv-001",
        "msg-001",
        "Hola, estoy pasando por un momento difícil y necesito guía espiritual"
    );

    await sleep(2000);

    // 2. Verificar que se guardó la memoria
    console.log("\n" + "=".repeat(60));
    console.log("2️⃣  VERIFICANDO MEMORIA GUARDADA");
    console.log("=".repeat(60));

    const sessions = await checkMemory(testUser);

    if (sessions === 0) {
        console.log("❌ No se guardó la memoria");
        return;
    }

    // 3. Segunda conversación (debería leer memoria)
    console.log("\n" + "=".repeat(60));
    console.log("3️⃣  SEGUNDA CONVERSACIÓN (con memoria)");
    console.log("=".repeat(60));

    await invokeLambda(
        testUser,
        "conv-002",
        "msg-002",
        "¿Qué versículo me recomiendas para encontrar paz?"
    );

    await sleep(2000);

    // 4. Ver logs
    console.log("\n" + "=".repeat(60));
    console.log("4️⃣  LOGS DEL LAMBDA");
    console.log("=".repeat(60));

    await getRecentLogs(15);

    // 5. Verificar memoria final
    console.log("\n" + "=".repeat(60));
    console.log("5️⃣  MEMORIA FINAL");
    console.log("=".repeat(60));

    await checkMemory(testUser);

    console.log("\n" + "=".repeat(60));
    console.log("✅ PRUEBA COMPLETADA");
    console.log("=".repeat(60));
    console.log(`\nUsuario de prueba: ${testUser}`);
    console.log("Revisa los logs para ver '✅ Found previous conversations' en la segunda llamada");
}

// Ejecutar
runTest().catch(error => {
    console.error("Error en la prueba:", error);
    process.exit(1);
});

